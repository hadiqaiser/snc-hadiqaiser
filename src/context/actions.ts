export enum UserAction {
  SetEnableLogs = "SET_ENABLE_LOGS",
  SetSelectedUser = "SET_SELECTED_USER",
  SetUserData = "SET_USER_DATA",
}
