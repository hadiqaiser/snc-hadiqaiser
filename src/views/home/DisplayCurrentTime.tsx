import {
  useEffect,
  FunctionComponent,
  PropsWithChildren,
  useMemo,
  useState,
} from "react";
import classNames from "classnames";

// Context
import { formatDateAndTime } from "@/utils/client";

// Font
import { Roboto_Mono } from "next/font/google";

const robotoMono = Roboto_Mono({ subsets: ["latin"] });

type DisplayCurrentTimeProps = {
  currentTime: Date;
  updateCurrentTime: (time: Date) => void;
};

export const DisplayCurrentTime: FunctionComponent<
  PropsWithChildren<DisplayCurrentTimeProps>
> = ({ currentTime, updateCurrentTime }) => {
  // States
  const [isMounted, setIsMounted] = useState(false);

  const formattedTime = useMemo(
    () => formatDateAndTime(currentTime),
    [currentTime],
  );

  useEffect(() => {
    setIsMounted(true);
    const interval = setInterval(() => {
      updateCurrentTime(new Date());
    }, 1000);

    return () => {
      clearInterval(interval);
    };
  }, [updateCurrentTime]);

  return (
    <p className={classNames(robotoMono.className, "leading-7")}>
      {isMounted && formattedTime}
    </p>
  );
};
