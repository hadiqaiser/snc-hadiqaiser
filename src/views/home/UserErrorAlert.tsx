import { FunctionComponent, PropsWithChildren } from "react";
import classNames from "classnames";
import { AlertCircle } from "lucide-react";

// UI
import { Alert, AlertDescription, AlertTitle } from "@/components/ui/alert";

type UserErrorAlertProps = {};

export const UserErrorAlert: FunctionComponent<
  PropsWithChildren<UserErrorAlertProps>
> = () => {
  return (
    <Alert variant="destructive" className={classNames("w-full sm:w-8/12")}>
      <AlertCircle className="h-4 w-4" />
      <AlertTitle>Error</AlertTitle>
      <AlertDescription data-testid="alert-description">
        Failed to fetch user information
      </AlertDescription>
    </Alert>
  );
};
