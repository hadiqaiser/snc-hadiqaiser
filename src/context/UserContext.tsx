import { createContext, ReactNode, useContext, useReducer } from "react";

// Actions
import { UserAction } from "./actions";

// Types
import { Person, User } from "@/utils/common/person";

type UserState = {
  enableLogs: boolean;
  selectedUser: Person | undefined;
  userData: User;
};

type UserContextValue = {
  state: UserState;
  dispatch: React.Dispatch<Action>;
};

type Action =
  | { type: UserAction.SetEnableLogs; payload: boolean }
  | { type: UserAction.SetSelectedUser; payload: Person | undefined }
  | { type: UserAction.SetUserData; payload: User };

const defaultValues: UserState = {
  enableLogs: false,
  selectedUser: undefined,
  userData: {
    backgroundImageUrl: "",
    profilePictureUrl: "",
    name: "",
    title: "",
    followers: 0,
    following: 0,
  },
};

interface Props {
  children: ReactNode;
}

export const UserContext = createContext<UserContextValue>({
  state: defaultValues,
  dispatch: () => null,
});

export const UserReducer = (state: UserState, action: Action): UserState => {
  switch (action.type) {
    case UserAction.SetEnableLogs:
      return { ...state, enableLogs: action.payload };
    case UserAction.SetSelectedUser:
      return { ...state, selectedUser: action.payload };
    case UserAction.SetUserData:
      return { ...state, userData: action.payload };
    default:
      return state;
  }
};

export const UserProvider: React.FC<Props> = ({ children }: Props) => {
  const [state, dispatch] = useReducer(UserReducer, defaultValues);

  return (
    <UserContext.Provider value={{ state, dispatch }}>
      {children}
    </UserContext.Provider>
  );
};

export const useUserContext = (): UserContextValue => {
  const safeContext = useContext(UserContext);
  if (!safeContext) {
    throw new Error("useUserContext must be used within a UserProvider");
  }

  return safeContext;
};
