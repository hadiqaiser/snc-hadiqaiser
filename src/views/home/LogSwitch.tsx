import { FunctionComponent, PropsWithChildren, useCallback } from "react";
import classNames from "classnames";

// UI
import { Label } from "@/components/ui/label";
import { Switch } from "@/components/ui/switch";

// Contexts
import { useUserContext } from "@/context/UserContext";
import { UserAction } from "@/context/actions";

type LogSwitchProps = {};

export const LogSwitch: FunctionComponent<
  PropsWithChildren<LogSwitchProps>
> = () => {
  // Context
  const {
    state: { enableLogs },
    dispatch,
  } = useUserContext();

  const toggleEnableLogs = useCallback(
    (checked: boolean) => {
      dispatch({
        type: UserAction.SetEnableLogs,
        payload: checked,
      });
    },
    [dispatch],
  );

  return (
    <div className={classNames("flex flex-row items-center space-x-2")}>
      <Switch
        id="toggle-logs"
        checked={enableLogs}
        onCheckedChange={toggleEnableLogs}
      />
      <Label htmlFor="toggle-logs">Enable Logs</Label>
    </div>
  );
};
