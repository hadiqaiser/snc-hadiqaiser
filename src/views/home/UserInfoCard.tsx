import { FunctionComponent, PropsWithChildren } from "react";
import classNames from "classnames";
import Image from "next/image";

// UI
import { Card } from "@/components/ui/card";
import { Separator } from "@/components/ui/separator";
import { Skeleton } from "@/components/ui/skeleton";

// Contexts
import { useUserContext } from "@/context/UserContext";

type UserInfoProps = {
  loading: boolean;
};

export const UserInfoCard: FunctionComponent<
  PropsWithChildren<UserInfoProps>
> = ({ loading }) => {
  // Hooks
  const {
    state: { userData },
  } = useUserContext();

  return (
    <Card className={classNames("relative w-full sm:w-8/12 h-auto")}>
      {loading ? (
        <Skeleton className={classNames("w-full h-52")} />
      ) : (
        <div className={classNames("relative w-full h-52")}>
          <Image
            src={userData.backgroundImageUrl || "/images/fallbackBgImg.svg"}
            alt="bg-pic"
            fill={true}
            className={classNames("rounded-md object-cover")}
          />
        </div>
      )}
      <div className={classNames("pb-6 px-6")}>
        <div
          className={classNames(
            "relative -mt-16 px-4 flex justify-start sm:-mt-20",
          )}
        >
          <div className={classNames("relative w-32 h-32 sm:w-40 sm:h-40")}>
            {loading ? (
              <Skeleton
                className={classNames("rounded-full w-32 h-32 sm:w-40 sm:h-40")}
              />
            ) : (
              <Image
                src={
                  userData.profilePictureUrl || "/images/fallbackProfileImg.svg"
                }
                alt="profile-pic"
                fill={true}
                className={classNames("rounded-full object-cover")}
              />
            )}
          </div>
        </div>
        <div
          className={classNames("flex flex-col sm:flex-row justify-between")}
        >
          <div className={classNames("flex flex-col gap-2 mt-3")}>
            {loading ? (
              <Skeleton className={classNames("h-8 w-40")} />
            ) : (
              <h3
                className={classNames(
                  "scroll-m-20 text-2xl font-semibold tracking-tight",
                )}
              >
                {userData.name || "No User"}
              </h3>
            )}
            {loading ? (
              <Skeleton className={classNames("h-7 w-48")} />
            ) : (
              <p className={classNames("text-xl text-muted-foreground")}>
                {userData.title}
              </p>
            )}
          </div>
          <div
            className={classNames("flex flex-row gap-5 mt-3 justify-evenly")}
          >
            <div
              className={classNames(
                "flex flex-col justify-between py-3 items-center",
              )}
            >
              {loading ? (
                <Skeleton className="h-6 w-12" />
              ) : (
                <p className={classNames("text-base text-muted-foreground")}>
                  {userData.followers}
                </p>
              )}
              <p className={classNames("text-base")}>Followers</p>
            </div>
            <Separator orientation="vertical" />
            <div
              className={classNames(
                "flex flex-col justify-between py-3 items-center",
              )}
            >
              {loading ? (
                <Skeleton className="h-6 w-12" />
              ) : (
                <p className={classNames("text-base text-muted-foreground")}>
                  {userData.following}
                </p>
              )}
              <p className={"text-base"}>Following</p>
            </div>
          </div>
        </div>
      </div>
    </Card>
  );
};
