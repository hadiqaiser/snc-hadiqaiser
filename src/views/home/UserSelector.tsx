import { FunctionComponent, PropsWithChildren, useCallback } from "react";
import classNames from "classnames";

// UI
import { Button } from "@/components/ui/button";

// Contexts
import { useUserContext } from "@/context/UserContext";
import { UserAction } from "@/context/actions";

// Common
import { Person } from "@/utils/common/person";

type UserSelectorProps = {};

export const UserSelector: FunctionComponent<
  PropsWithChildren<UserSelectorProps>
> = () => {
  // Hooks
  const {
    state: { selectedUser },
    dispatch,
  } = useUserContext();

  const handleUserChange = useCallback(
    (person: Person) => {
      dispatch({
        type: UserAction.SetSelectedUser,
        payload: person,
      });
    },
    [dispatch],
  );

  return (
    <div
      className={classNames(
        "w-full sm:w-8/12 flex flex-col sm:flex-row gap-4 justify-evenly mt-8",
      )}
    >
      {Object.values(Person).map((person) => (
        <Button
          variant={"outline"}
          key={person}
          onClick={() => handleUserChange(person)}
          className={classNames({ "border-black": selectedUser === person })}
        >
          {person}
        </Button>
      ))}
    </div>
  );
};
