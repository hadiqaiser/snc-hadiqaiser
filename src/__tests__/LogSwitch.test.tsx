import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";

// Views
import { LogSwitch } from "@/views/home/LogSwitch";

// Context
import { useUserContext } from "@/context/UserContext";
import { UserAction } from "@/context/actions";

jest.mock("../context/UserContext", () => ({
  useUserContext: jest.fn(),
}));

describe("Log Switch", () => {
  const mockDispatch = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    (useUserContext as jest.Mock).mockReturnValue({
      state: { enableLogs: false },
      dispatch: mockDispatch,
    });
  });
  it("should match the LogSwitch snapshot", () => {
    const testRender = render(<LogSwitch />);

    expect(testRender).toMatchSnapshot();
  });

  it("should render correctly with initial state", () => {
    render(<LogSwitch />);

    const switchElement = screen.getByRole("switch");
    const labelElement = screen.getByText("Enable Logs");

    expect(switchElement).toBeInTheDocument();
    expect(labelElement).toBeInTheDocument();
    expect(switchElement).not.toBeChecked();
  });

  it("should call dispatch with correct action when toggled", () => {
    render(<LogSwitch />);

    let switchElement = screen.getByRole("switch");

    fireEvent.click(switchElement);

    expect(mockDispatch).toHaveBeenCalledWith({
      type: UserAction.SetEnableLogs,
      payload: true,
    });
  });
});
