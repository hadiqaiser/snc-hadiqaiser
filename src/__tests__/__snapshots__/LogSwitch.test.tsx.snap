// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log Switch should match the LogSwitch snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-row items-center space-x-2"
      >
        <button
          aria-checked="false"
          class="peer inline-flex h-6 w-11 shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=unchecked]:bg-input"
          data-state="unchecked"
          id="toggle-logs"
          role="switch"
          type="button"
          value="on"
        >
          <span
            class="pointer-events-none block h-5 w-5 rounded-full bg-background shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-5 data-[state=unchecked]:translate-x-0"
            data-state="unchecked"
          />
        </button>
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
          for="toggle-logs"
        >
          Enable Logs
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-row items-center space-x-2"
    >
      <button
        aria-checked="false"
        class="peer inline-flex h-6 w-11 shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=unchecked]:bg-input"
        data-state="unchecked"
        id="toggle-logs"
        role="switch"
        type="button"
        value="on"
      >
        <span
          class="pointer-events-none block h-5 w-5 rounded-full bg-background shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-5 data-[state=unchecked]:translate-x-0"
          data-state="unchecked"
        />
      </button>
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for="toggle-logs"
      >
        Enable Logs
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
